### ISSUE TITLE
Unexpected behavior for string input with high precision

### ISSUE THREAD

**Issue by josdejong** [rawify/Fraction.js#28]
A user of math.js noticed some unexpected behavior when creating a `Fraction`:

```js
// works as expected:
new Fraction(4.166666666666667) // {s: 1, n: 25, d: 6}

// doesn't work as expected (the second value was coming from a BigNumber)
new Fraction('4.166666666666667') 
    // {s: 1, n: 4166666666666667, d: 1000000000000000}
new Fraction('4.166666666666666666666666666666666666666666666666666666666666667') 
    // {s: 1, n: 7602530730928912, d: 1824607375422939}
```

I guess this has something to do with round-off errors and the stop condition of the loop that tries to split the value into a fraction, but I don't really understand it. Do you have any ideas why this happens?

**Comment by infusion** [comment]
Ah and BTW, the actual problem I have with BigInt is backwards compatibility. Every constant term within the source must be suffixed by n, like "0n" for a BigInt(0). This is cool since normal operators are overloaded to work on BigInt and so only few changes on operations must be done. However, I don't see a way to make it work on old JS installations without breaking the source into two distributions. Any plans on how you will handle that on the math.js side?

**Comment by infusion** [comment]
Yea, lets focus on fractions first, complex numbers is another even more complex story. I'm almost done converting it to bigint, from 200 failing tests to only 29 edge cases. I think I will deliver the file as a second source file for a while, until support on browsers is better, what do you think? On math.js you can explicitly link to that file then.

**Comment by infusion** [comment]
Boooooom, I invite all of you to test the file as much as possible: https://github.com/infusion/Fraction.js/blob/master/bigfraction.js

Only one test is failing, namely `JSON.stringify` does not know how to serialize BigInt, I expect that to be fixed in future versions of v8.

**Comment by infusion** [comment]
Ah and what I changed is this silly `NaN` behavior. Since `NaN ∉ BigInt`, I reverted that decision that `NaN ∈ Fraction`. Based on a pull request somewhere in time we got `Fraction(NaN)='NaN'`. Now it is `Fraction(NaN)=throw invalid param`. This cleans the code a lot and a real `NaN` was rare anyways since parsers were throwing on invalid params already. Only `NaN` was a special case. And we somehow had a wrong test case with a 1500 digits repeating part in the test suite. BigInt now has the precision to cover that case correctly.

**Comment by infusion** [comment]
And to come back to the initial call of issue #28:

```js
new Fraction('4.166666666666667')
// { s: 1n, n: 4166666666666667n, d: 1000000000000000n }
new Fraction('4.166666666666666666666666666666666666666666666666666666666666667')
// { s: 1n,
//   n:
//    4166666666666666666666666666666666666666666666666666666666666667n,
//   d:
//    1000000000000000000000000000000000000000000000000000000000000000n }
```

**Comment by infusion** [comment]
I just managed to implement a backwards compatible fraction.js implementation. The only problem is Babel now. I got a ticket a while ago that es6 modules can't be compressed easily with Babel. Does anyone know details on that so that we can make bigfraction.js the new fraction.js file?

**Comment by josdejong** [comment]
That sounds really promising Robert!!!

I regularly have difficulties with Webpack too. Maybe you can find an other library which has a similar ES6 setup as you want for Fraction.js has, to see how it's done.

**Comment by infusion** [comment]
I hope we can find a solution for the packaging problem as it is the only show stopper right now to release 5.0.0. Do you see any conflicts on the math.js side when the attributes {s,n,d} are bigints? I was rowing back from ES6 rapidly when I got the ticket that this change breaks webpack/babel setups and I couldn't find other projects implementing it cleanly. @josdejong do you know someone in the webpack universe to solve this issue for fraction.js as well as math.js?

**Comment by josdejong** [comment]
It's not entirely clear what the Webpack issue is you're encountering. the math.js code base is now ES6 (imports, export, const, let, etc.), and all works fine including minification.

If the BigInt is only used internally in Fraction.js we could consider it not to be a breaking change. But even if it is a breaking change it's no problem to do a major release. From math.js point of view it will be easiest then to schedule it for the upcoming v6.0.0.

**Comment by infusion** [comment]
It might only break when you access the numerator and denominator attributes directly since you run into a type mismatch this way. The quick fix is to wrap it into a `Number(x)` cast, the harder might be to use the BigInts correctly everywhere.
We can also say to publish fraction.js 5.0.0 in ES6 and when things break again for some people, fix it when the problem occurs. 
For when do you plan to release 6.0.0?

**Comment by josdejong** [comment]
> We can also say to publish fraction.js 5.0.0 in ES6 and when things break again for some people, fix it when the problem occurs.

Do you mean like the library only offers ES6 code and no version transpiled into ES5? I think that will be problematic for all uses of the library isn't it?

> For when do you plan to release 6.0.0?

That will take quite some time, I guess at least a month. Currently the middle of a big refactoring to make the library tree-shake-friendly.

**Comment by josdejong** [comment]
@infusion I would try to find a library or tutorial which has a setup that appeals to you, go from there having an empty but working setup, and then put the code of Fraction.js in place there until everything works.

**Comment by m93a** [comment]
@infusion @josdejong I summed up math.js's current capabilities of the representation of exact numbers in a [short essay](https://github.com/m93a/mathjs-sets/blob/master/README.md). The real-life consequences of that analysis are minimal, one rarely needs more than even 32bit floats, let alone exact representation, but you still might find it interesting.

**Comment by josdejong** [comment]
:+1:

**Comment by josdejong** [comment]
It would be interesting if we could inject a custom data type into Fraction.js, so you can use `BigInt` instead of `number`. Similar to ideas I've shared in the past for Complex.js: would be cool if you could use say BigNumbers there.

**Comment by josdejong** [comment]
For reference: https://github.com/josdejong/mathjs/issues/694#issuecomment-235680880

**Comment by infusion** [comment]
@josdejong I'm not in favor of introducing such an abstraction layer to fraction.js. However, I implemented a BigInt variation of fraction.js quite a while ago in file `bigfraction.js`, which runs all unit tests perfectly, but sets all operations on ES6 native BigInt implementation. I would go this route rather than implementing a switch, like I did in Polynomial.js to keep it as simple as possible.

**Comment by josdejong** [comment]
Ok clear, thanks for your explanation.

**Comment by infusion** [comment]
So I would go with the BigInt version completely in the future. But the question for you is what backward compatibility you want. I don't know how BigInt support is on Babel.

**Comment by josdejong** [comment]
Yes indeed, makes sense to move to BigInt over time.

**Comment by infusion** [comment]
You can of course make first tests integrating the test file bigfraction.js. Firefox supports BigInt since v68, Chrome since v67 and Edge since v79, with about 70% of the internet community according to https://caniuse.com/#feat=bigint

**Comment by josdejong** [comment]
Yes that's true, all major browsers currently support BigInt already! That's moving faster than I expected.

**Comment by infusion** [comment]
Are there stats for nodejs versions? I would assume that node versions are kept quite up to date and V8 introduced this feature back in 2018. I think the only problem arises when people access `n` and `d` attributes directly, for all other things bigfraction.js can serve as a drop-in replacement of the original fraction.js.

**Comment by m93a** [comment]
Bigints can be polyfilled [if we use a wrapper](https://github.com/peterolson/BigInteger.js/). This would let us use _native_ arbitrary precision arithmetic with modern browsers, but it would still work in old browsers.

**Comment by a-zhendi** [comment]
Hi @infusion! I'm trying to use Math.js + bigfraction.js (via webpack alias `'fraction.js': 'fraction.js/bigfraction.js'`). It seems to work fine. The problem (in my case) is that BigInt is not supported in Safari and iOS Safari yet.

After some research I've found [JSBI](https://github.com/GoogleChromeLabs/jsbi) from GoogleChromeLabs. It is a pure-JavaScript implementation of the official ECMAScript BigInt proposal.
[Because of impossibility to polyfill BigInt directly](https://github.com/GoogleChromeLabs/jsbi#why) they propose a different strategy:

1. Base your code on JSBI instead of native BigInt.
2. [Transpile it to native BigInt](https://github.com/GoogleChromeLabs/babel-plugin-transform-jsbi-to-bigint) when available.

What you think about it?
What you think about refactoring `bigfraction.js` to `JSBI` until native BigInt will get wider support? Or may be creating additional `bigfraction-jsbi.js` version?

**Comment by m93a** [comment]
Hey @a-zhendi,
My thinking was that the ideal solution would be a polyfill. Yes, it is impossible to polyfill BigInt syntax, but it's still possible to use the native BigInt *implementation* whenever it's available. JSBI doesn't do this, instead it uses its own implementation no matter what. The library I posted a link to, [BigInteger.js](https://github.com/peterolson/BigInteger.js/), does exactly what I described, but the implementation it uses when native BigInt isn't available might be slower than JSBI.

To really polyfill BigInts we could either a) use BigInteger.js, or b) detect whether the browser supports BigInt, if not use JSBI, else use native bigint.

I might do a performance test later today, so that we see how the variants compare. We might learn that there's no real performance benefit in using native BigInts, then it would be reasonable to use just JSBI, without any polyfills.
